routing_plugin.go

```diff
@@ -107,3 +120,518 @@ func ReleaseRoutingPlugin(ctx context.Context) {
 		routingPluginInstance = nil
 	}
 }
+
+// lowToHighRoutingPlugin is a RoutingPlugin that implements "low to high"
+// routing. This means that when we're attempting to pay to a target we'll
+// gradually (with a linear step function) discard inbound peers to that target
+// given routing timeouts. The lowToHighRoutingPlugin itself is responsible for
+// manipulating LND's Mission Control to make such routing attempts possible.
+type lowToHighRoutingPlugin struct {
+	lnd     lndclient.LndServices
+	clock   clock.Clock
+	target  route.Vertex
+	amount  btcutil.Amount
+	mcState map[route.Vertex]lndclient.MissionControlEntry
+
+	// nodesByMaxFee holds nodes sorted by maximum fees that would be paid
+	// to the target node for the target amount.
+	nodesByMaxFee []nodeFeeInfo
+
+	// mcChanged flags that the MC settings for the tracked nodes were
+	// changed and should be reset to their original state once the plugin
+	// is done.
+	mcChanged bool
+}
+
+type nodeFeeInfo struct {
+	node     route.Vertex
+	capacity btcutil.Amount
+	fee      int64
+}
+
+// Enforce that lowToHighRoutingPlugin implements the RoutingPlugin interface.
+var _ RoutingPlugin = (*lowToHighRoutingPlugin)(nil)
+
+// buildPrivateChannels creates the private channel map from the passed route
+// hints. The code is taken and adapted from LND. Original source:
+// lnd/routing/payment_session_source.go.
+func buildPrivateChannels(routeHints [][]zpay32.HopHint,
+	target route.Vertex) map[route.Vertex][]*lndclient.ChannelEdge {
+
+	edges := make(map[route.Vertex][]*lndclient.ChannelEdge)
+
+	// Traverse through all of the available hop hints and include them in
+	// our edges map.
+	for _, routeHint := range routeHints {
+		// If multiple hop hints are provided within a single route
+		// hint, we'll assume they must be chained together and sorted
+		// in forward order in order to reach the target successfully.
```

Comment:
By chained together do you mean this layout?
```
Client ---- N1 ---(private)--- Gateway  ---(private)--- Server
```

I see that our docs currently say "chained together to make a route", but I'm pretty sure we don't support beyond depth 1 hints. If we have multiple hints it'll be this sort of setup:
```
                     ---(private)--- 
Client ----- Gateway ---(private)--- Server
                     ---(private)--- 
```